packages/dropdownable/components/Dropdownable.tsx

```diff
@@ -0,0 +1,246 @@
+}
+
+interface PositionCoord {
+  top: number;
+  left: number;
+}
+
+interface State {
+  direction: Direction;
+  position: PositionCoord;
+}
+
+const METHODS_TO_BIND = [
+  "getDropdown",
+  "setPosition",
+  "setPositionFromCurrentProps",
+  "calculateBestDirection",
+  "positionForDirection",
+  "childBounds",
+  "windowDimensions",
+  "dropdownDimensions"
+];
+
+class Dropdownable extends React.Component<DropdownableProps, State> {
+  public static defaultProps: Partial<DropdownableProps> = {
+    preferredDirections: DEFAULT_DIRECTION_PREFERENCES
+  };
+
+  private child = React.createRef<HTMLDivElement>();
+  private dropdown = React.createRef<HTMLDivElement>();
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      direction: Direction.BottomLeft,
+      position: {
+        top: 0,
+        left: 0
+      }
+    };
+
+    METHODS_TO_BIND.forEach(method => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentDidMount() {
+    resizeEventManager.add(this.setPositionFromCurrentProps);
+    this.setPositionFromCurrentProps();
+  }
+
+  componentWillUnmount() {
+    resizeEventManager.remove(this.setPositionFromCurrentProps);
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.open) {
+      this.setPosition(nextProps);
+    }
+  }
+
+  render() {
+    const Container = styled("div")`
+      position: relative;
+      display: inline-block;
+    `;
+
+    return (
+      <Container>
+        <div ref={this.child}>{this.props.children}</div>
+        {this.getOverlay()}
+      </Container>
+    );
+  }
+
+  getOverlay() {
+    const { position } = this.state;
+    const { open } = this.props;
+
+    const overlayStyle = css({
+      top: `${position.top}px`,
+      left: `${position.left}px`,
+      position: "absolute",
+      transform: `scale(${open ? 1 : 0})`,
+      opacity: open ? 1 : 0
+    });
+
+    return <Overlay className={overlayStyle}>{this.getDropdown()}</Overlay>;
+  }
+
+  getDropdown() {
+    const { open, dropdown, onClose } = this.props;
+
+    return (
+      <div ref={this.dropdown}>
+        <DropdownContents open={open} onClose={onClose}>
+          {dropdown}
+        </DropdownContents>
+      </div>
+    );
+  }
+
+  setPositionFromCurrentProps() {
+    this.setPosition(this.props);
+  }
+
+  setPosition(props: DropdownableProps) {
+    const dropdownDimensions = this.dropdownDimensions();
+    const windowDimensions = this.windowDimensions();
+    const childBounds = this.childBounds();
+    const direction = this.calculateBestDirection(
+      props.preferredDirections,
+      childBounds,
+      dropdownDimensions,
+      windowDimensions
+    );
+
+    this.setState({
+      direction,
+      position: this.positionForDirection(
+        direction,
+        childBounds,
+        dropdownDimensions
+      )
+    });
+  }
+
+  calculateBestDirection(
+    preferredDirections,
+    childBounds,
+    dropdownDimensions,
+    windowDimensions
```

Comment:
It would be helpful to type these params as well as `positionForDirection` params, and `childBounds`, `windowBounds`, and `dropdownDimensions` return types